### PR TITLE
[cmake] Auto-complete via clangd auto-setup v2

### PR DESCRIPTION
**Pull request by ntrrgc**
#### 4b85cbe26eff585fe8586af43bd7199f326b0caf
<pre>
[cmake] Auto-complete via clangd auto-setup v2
<a href="https://bugs.webkit.org/show_bug.cgi?id=274909">https://bugs.webkit.org/show_bug.cgi?id=274909</a>

Reviewed by Elliott Williams.

Re-upload of the patch including a fix for the internal Apple builds
that do not contain a Tools/ directory. See
<a href="https://bugs.webkit.org/show_bug.cgi?id=274909#c7">https://bugs.webkit.org/show_bug.cgi?id=274909#c7</a> for details.

Additionally, this version has been modified to not depend on PyYAML or
Jinja2 and use simpler stdlib replacements instead, as Apple internal
builds cannot use any non-builtin Python packages.

---

 # Goals

This patch aims to bring out-of-the box completion with clangd for
WebKit developers using CMake builds.

clangd is an open source language server for providing code-completion,
go-to-definition and similar smart features for C++ code by analyzing it
with the same backend clang uses. **You do not need to compile WebKit
with clang to take advantage of this**, as clang strives to be flag
compatible with GCC, and even MSVC through `clang-cl`.

Using clangd with WebKit was already possible before, but required
manual intervention in the project files to set it up.

This patch attempts to clear some of the existing pitfalls,
particularly for CMake builds, so that more people take advantage of
Clangd without having to put conscious effort into troubleshooting.

This patch also makes WebKit stock clangd setup compatible with both
Mac and non-Mac platforms and allows developers to customize clangd
configuration without dirtying their repository.

---

Getting clangd right takes a lot of context and moving pieces, so this
is one of those very long commit messages. Sections with (NEW) in the
title explain changes introduced by this patch, whereas the rest hope to
explain the background and why those changes are necessary or useful.

 # Architecture (NEW)

Several scripts have been added in Tools/clangd. A diagram describing
the build system targets, dependencies and products is also included.

The clangd auto-setup needs to generate the following artifacts:
 * .clangd configuration file
 * compile_commands.json file/symlink in the WebKit repo root.

Starting with this patch `.clangd` is no longer committed in the
repository it is now .gitignore&apos;d and auto-generated during the build.

 # Compatibility with XCode builds (NEW)

This patch doesn&apos;t improve or worsen the status of clangd support in
XCode builds.

XCode build users are still expected to generate compile_commands.json
manually as per the wiki states:
<a href="https://trac.webkit.org/wiki/Clangd#macOS1">https://trac.webkit.org/wiki/Clangd#macOS1</a>

    make release EXPORT_COMPILE_COMMANDS=YES
    generate-compile-commands WebKitBuild/Release

The generation of the `.clangd` file is also included in the XCode
build, as part of the WTF XCode project.

 # clangd config generation (NEW)

The script `update-clangd-config` autogenerates a `.clangd` file with an
«autogenerated» marker at the top. The file will only be updated as long
as that marker exists, therefore allowing end-user customization if
needed.

The `.clangd` file is generated from a template, which will produce
slightly different options for Mac and non-Mac platforms, fixing the
regression introduced in <a href="https://github.com/WebKit/WebKit/pull/23273.">https://github.com/WebKit/WebKit/pull/23273.</a>
The template previously used Jinja2, but in this version of the patch
uses the more rudimentary stdlib&apos;s strings.Template to avoid depending
on external libraries.

`update-clangd-config` is made part of the build by default. For XCode
builds, it is generated as part of WTF. For CMake builds, it is enabled
as a default target as long as CLANGD_AUTO_SETUP is set, which is the
default when building inside the WebKitBuild directory.

 # clangd configuration changes (NEW)

This patch also makes several changes to the clangd configuration file
to fix issues detected during testing:

 * UnifiedSource files are no longer indexed.
 * The `-fcoroutines` flag is stripped, as it is not recognized by some
   versions of clang, and by extension, clangd.
 * `-xobjective-c++` is now only passed in macOS to avoid depending on
   macOS-only headers in other systems.
 * `config.h` is no longer prepended to files within the ThirdParty
   directory.

 # compile_commands.json generation: context

CMake already generates a compile_commands.json file when the Ninja or
Makefile file is generated as long as CMAKE_EXPORT_COMPILE_COMMANDS is
ON, which is already the default in DEVELOPER_MODE.

When using unified builds (which is the default), that
`compile_commands.json` is somewhat subpar, as the language server is
unaware of most of the source .cpp files and hence has to require to
heuristics to match them to other build commands.

On the CMake side, a RewriteCompileCommands target exists that parses
that file along with unified sources to generate an improved file at
`&lt;BUILD_DIR&gt;/DeveloperTools/compile_commands.json`. Unfortunately, this
may not be obvious to many developers, which may be unaware of its
existence.

On the XCode builds side, this responsibility would fall under the
`generate-compile-commands` script, but it&apos;s seemingly not currently
implemented. Fixing that shouldn&apos;t be a difficult change but it&apos;s
outside of the scope of this patch. For more details:
<a href="https://github.com/WebKit/WebKit/pull/30784#issuecomment-2264933253">https://github.com/WebKit/WebKit/pull/30784#issuecomment-2264933253</a>

 # Same compile_commands.json path for non-unified builds (NEW)

The first change this patch makes is solving the unified builds vs
non-unified builds dichotomy by introducing a new build target for
non-unified builds named SymlinkCompileCommandsInDeveloperTools.

This way, **no matter the type of build**,
`&lt;BUILD_DIR&gt;/DeveloperTools/compile_commands.json` is the path to be
relied upon.

 # compile_commands.json vs builds

For a `compile_commands.json` file to be used, it needs to exist in a
parent directory of the sources being worked on.

Since any generated `compile_commands.json` is specific for a given build
(e.g. Debug vs Release), with build-specific flags, developers need to
make a choice of what build to use for the `compile_commands.json` in
their WebKit checkout.

In XCode builds, this is passed explicitly to the
generate-compile-commands command.

In CMake builds, this is usually accomplished by symlinking the file
into the root of the WebKit repository.

Developers who frequently switch and remove builds need to be careful to
not let the compile_commands.json symlink become broken, as that would
cause the language server to rely on much worsened heuristics.

 # compile_commands.json symlink autogeneration and management (NEW)

This patch introduces automatic compile_commands.json symlink handling.
A new target is introduced that intentionally runs with every build:
`UpdateCompileCommandsSymlink`, which relies on the
`update-compile-commands-symlink` script, also introduced by this patch.
The script is meant to be fast enough to not introduce any perceivable
delay in builds.

`update-compile-commands-symlink` receives as arguments a symlink path
and a configuration file listing a series of candidate paths. The
symlink is updated to point to the first candidate path that exists.
In the previous version of this patch, that file was in YAML format, but
this version uses Python&apos;s ConfigParse INI syntax to avoid depending on
any external libraries.

This list of candidate paths is declared in a new .gitignore&apos;d
`update-compile-commands-symlink.conf` file that is generated early
during any CMake build that uses `CLANGD_AUTO_SETUP`.

`update-compile-commands-symlink.conf` will be automatically updated
from `Tools/clangd/update-compile-commands-symlink.conf.example`; but
similarly to `.clangd`, a user can disable this and manage the file on
their own, in this case by setting the `user_managed` property inside
the file to True.

 # Testing

clangd auto-setup has been tested with a Linux x86_64 machine using
webkit-container-sdk and Visual Studio Code, both with the default GCC
and with clang 16, with a fresh installation of VS Code (no settings).

It is possible to confirm if a file passes under clangd by using the
`clangd --check` command along with the compile lines used during
compilation, which can be queried from compile_commands.json.

A couple barebones scripts showing this usage have been provided in
Tools/clangd as documentation: `clangd-check-file` and
`clangd-check-codebase`.

I was able to run clangd-check-codebase with very few .cpp files
failing, and the failing files had good reasons for failing, like
missing includes that were obscured by unified builds.

Canonical link: <a href="https://commits.webkit.org/282175@main">https://commits.webkit.org/282175@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6f61716370020b70f34faafb74820b556d3feb8a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64823 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49234 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7937 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30060 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10352 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53994 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55985 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10277 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66552 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60139 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10102 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56790 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13850 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4009 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81894 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36056 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14271 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37138 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38231 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36883 "Built successfully") | | |

### CODE REVIEW

**Comment by ntrrgc** [comment]
@emw-apple Thanks for the snippet in https://bugs.webkit.org/show_bug.cgi?id=274909#c7 This is a re-upload with that applied.

```diff
diff --git a/Source/WTF/WTF.xcodeproj/project.pbxproj b/Source/WTF/WTF.xcodeproj/project.pbxproj
index 8568a9bf3be9..f3225446e906 100644
--- a/Source/WTF/WTF.xcodeproj/project.pbxproj
+++ b/Source/WTF/WTF.xcodeproj/project.pbxproj
@@ -3876,7 +3876,7 @@
                        );
                        runOnlyForDeploymentPostprocessing = 0;
                        shellPath = /bin/sh;
-                       shellScript = "\"${SCRIPT_INPUT_FILE_0}\" \"${SCRIPT_INPUT_FILE_1}\" \"${SCRIPT_OUTPUT_FILE_0}\"\n";
+                       shellScript = "if [ \"${DEPLOYMENT_LOCATION}\" != YES ]; then\n    \"${SCRIPT_INPUT_FILE_0}\" \"${SCRIPT_INPUT_FILE_1}\" \"${SCRIPT_OUTPUT_FILE_0}\"\nfi\n";
                };
                DD07DF4C2A7DB6640091EACE /* Symlink public SDK headers */ = {
                        isa = PBXShellScriptBuildPhase;
```

**Review by emw-apple — CHANGES_REQUESTED** [review]
Thanks for the ping!

`update-clangd-config` using an auto-installed jinja2 will unfortunately be an issue for us, as we need to be able to build WebKit in environments where there is no network access.

One thing I'm thinking about is, does the clangd generator really need to be this complex?  If I understand correctly, the only thing that is dynamically generated is the `-xobjective-c++` argument when building on a Mac.

What if we had multiple clangd files (`Tools/clangd/clangd-Darwin.yaml`, `Tools/clangd/clangd-Linux.yaml`, etc.), and all the build-time script did was symlink the correct one into the top-level? It could even be done with a shell one-liner like:

    ln -sfh Tools/clang/clangd-$(uname -s).yaml "${WK_WORKSPACE_DIR}/.clang"

**Comment by emw-apple** [comment]
Another idea would be to keep the templating, but use Python's built-in [`string.Template`](https://docs.python.org/3/library/string.html#template-strings) library instead of jinja, so that the generator script can run (quickly!) in firewalled build environments.

**Comment by ntrrgc** [comment]
@emw-apple Is PyYAML similarly a problem for you as is Jinja2?

**Comment by emw-apple** [comment]
> @emw-apple Is PyYAML similarly a problem for you as is Jinja2?

Unfortunately yes. Our build scripts need to run using standard library Python only.

**Comment by emw-apple** [comment]
Unfortunately yes, we only have the Python standard library in our toolchains. Though we do support [`yaml`](https://ruby-doc.org/stdlib-2.7.0/libdoc/yaml/rdoc/YAML.html) in Ruby and [`YAML`](https://metacpan.org/pod/YAML) in Perl, if you're open to writing the build script in another language.

**Comment by tuankiet65** [comment]
I was going to modify `.clangd` to add a flag, but since you're working to overhaul clangd support, I might as well let you know about it. I don't know if you've run into it, but my clangd setup (clangd bundled with Xcode) would emit this error to every opened files: "Too many errors emitted, stopping now". Turns out, it's possible when clangd processes a file, that clang emits a bunch of warnings, more specifically unused functions in other files. clangd filters those out, but since we build with `-Werror`, those warnings get turned into errors, and after 20 errors (the default limit), clang would emit that "too many errors emitted..." error, which clangd then presents to the user. I fix this by adding [`-ferror-limit=0`](https://clang.llvm.org/docs/UsersManual.html#cmdoption-ferror-limit) to disable the error limit. I think we should add this flag to the `.clangd` template

**Comment by ntrrgc** [comment]
@tuankiet65 What build (port, flags, etc.) are you experiencing that problem in? I don't understand why clangd would complain about unused functions in other files.

**Comment by ntrrgc** [comment]
@tuankiet65 But if the file is normally built with `-Werror`, how come it compiles? Normally the intention is that clangd flags should match compiler flags where possible, so an error in one should be an error in the other and vice versa.

**Comment by tuankiet65** [comment]
@ntrrgc yeah that's one of the mysteries I haven't figured out. If I manually executed the compile commands for that file, it compiled fine with no errors, but somehow it throws errors when used with clangd. Maybe there're differences between clangd using clang vs. using clang directly, but I don't know

**Comment by ntrrgc** [comment]
@tuankiet65 I just built it macOS, both the test the new iteration of the patch I'm about to upload and to have a look at the problem you described. **I was able to reproduce the "too many errors" problem**. I identified several potential problems in the process:

1. The `generate-compile-commands` script used in XCode builds **doesn't generate entries for .cpp files that are part of an UnifiedBuilds file**. When this is the case, clangd has to guess flags from another .cpp file that is in `compile_commands.json`, and this can go wrong. For instance, `CSSRayValue.cpp` was using flags from CDM.cpp from encryptedmedia.

   In CMake builds we have a script to generate entries for those files: `rewrite-compile-commands`. It is currently unused in XCode builds and it would need some minor tweaks (attached below [1]), but it would be easy to add.

   In the end, this wasn't the source of the problem though.

2. Many internal functions in bmalloc are annotated with **`PAS_UNUSED`** which supresses warnings like that. There are a few that aren't. `bmalloc_allocate_impl_casual_case` is one of them, and adding `PAS_UNUSED` indeed makes the warning for that one go away.

3. These unused function warnings are only visible with `--log=verbose` and they are **"dropped diagnostics"**. I wonder if those should count towards the error limit and if them counting in clangd is a bug.

At this point, I would be getting too sidetracked if I continue to dig, and therefore I find **`"-ferror-limit=0"` a reasonable solution** for the time being, which will add to the patch. If anyone wants to pursue any of the issues described above, please go ahead with them, but they are not blockers for this patch.

[1] Small patch for making `rewrite-compile-commands` support the `compile_commands.json` files created by Xcode:
```diff
diff --git a/Tools/Scripts/rewrite-compile-commands b/Tools/Scripts/rewrite-compile-commands
index 8109574fabbd..7fb50116c133 100755
--- a/Tools/Scripts/rewrite-compile-commands
+++ b/Tools/Scripts/rewrite-compile-commands
@@ -79,9 +79,13 @@ for entry in compile_commands:
                 include_file = os.path.join(d, include_path)

                 if os.path.exists(include_file):
+                    if "command" in entry:
+                        command = entry["command"]
+                    else:
+                        command = " ".join(entry["arguments"])
                     generated_compile_commands.append({
                         "directory": entry["directory"],
-                        "command": entry["command"].replace(entry_file, include_file),
+                        "command": command.replace(entry_file, include_file),
                         "file": include_file,
                     })
                     break
```

**Comment by ntrrgc** [comment]
@emw-apple Please check if there remain any issues with the updated patch.

**Comment by tuankiet65** [comment]
@ntrrgc thank you for the throughout investigation!

> The generate-compile-commands script used in XCode builds doesn't generate entries for .cpp files that are part of an UnifiedBuilds file. When this is the case, clangd has to guess flags from another .cpp file that is in compile_commands.json, and this can go wrong. For instance, CSSRayValue.cpp was using flags from CDM.cpp from encryptedmedia.
> 
> In CMake builds we have a script to generate entries for those files: rewrite-compile-commands. It is currently unused in XCode builds and it would need some minor tweaks (attached below [1]), but it would be easy to add.
> 
> In the end, this wasn't the source of the problem though.
 
Yeah, I ran into the same trouble using `rewrite-compile-commands` with the `compile_commands.json` generated from Xcode builds. I had a similar patch to you, albeit different approach, since your way of concatenating arguments scared me a bit. If you're willing to incorporate this into the PR, that'd be good, else I'll do it on a separate PR.

```diff
diff --git a/Tools/Scripts/rewrite-compile-commands b/Tools/Scripts/rewrite-compile-commands
index 8109574fabbd..1eb16faacba0 100755
--- a/Tools/Scripts/rewrite-compile-commands
+++ b/Tools/Scripts/rewrite-compile-commands
@@ -79,12 +79,18 @@ for entry in compile_commands:
                 include_file = os.path.join(d, include_path)

                 if os.path.exists(include_file):
-                    generated_compile_commands.append({
+                    new_entry = {
                         "directory": entry["directory"],
-                        "command": entry["command"].replace(entry_file, include_file),
                         "file": include_file,
-                    })
-                    break
+                    }
+
+                    if "command" in entry:
+                        new_entry["command"] = entry["command"].replace(entry_file, include_file)
+                    else:
+                        arguments = [arg.replace(entry_file, include_file) for arg in entry["arguments"]]
+                        new_entry["arguments"] = arguments
+
+                    generated_compile_commands.append(new_entry)

 output_directory = os.path.dirname(args.output_file)
 if not os.path.isdir(output_directory):
```

> Many internal functions in bmalloc are annotated with PAS_UNUSED which supresses warnings like that. There are a few that aren't. bmalloc_allocate_impl_casual_case is one of them, and adding PAS_UNUSED indeed makes the warning for that one go away.

Maybe it's best for us to add `PAS_UNUSED` to the unused function, but sounds like a problem for future me.

> These unused function warnings are only visible with --log=verbose and they are "dropped diagnostics". I wonder if those should count towards the error limit and if them counting in clangd is a bug.

I guess clang produces those diagnostics and counts the errors to produce the "too many errors" error. Then clangd dropped those diagnostics, but it didn't know the "too many errors" error stems from clang counting dropped diagnostics as errors.

> At this point, I would be getting too sidetracked if I continue to dig, and therefore I find "-ferror-limit=0" a reasonable solution for the time being, which will add to the patch

If you don't want to add it (for it being a hack), I could try to get the unused functions marked as `PAS_UNUSED`, that way we wouldn't have to use `-ferror-limit=0`. I'm okay with either approach.

**Comment by ntrrgc** [comment]
> your way of concatenating arguments scared me a bit

That is fair. The compile_commands.json CMake generates use `"command"` instead of `"arguments"` (unfortunately CMake handles spaces in paths very poorly) so it was easier to do it that way, but doing it on `"arguments"` when available is a better solution.

> If you don't want to add it [`-ferror-limit=0`]

I have already added it to the version of the patch currently in review, as I mentioned earlier.

File: .gitignore
Comment (emw-apple) [inline .gitignore:8]: Can this go in a `Tools/clangd/.gitignore` file? Doesn't seem like it needs to be stored in the top level directory.

**Review by emw-apple — APPROVED** [review]
The bits that touch the Xcode build look fine to me. I'd recommend getting a second review from someone who knows the Linux + CMake build better.

File: .gitignore
Comment (ntrrgc) [reply .gitignore:8]: As far I know, not really, as those ignore patterns would only apply inside `Tools/clangd`.

I tested doing the change just in case, and it indeed seem to only apply to `Tools/clangd` and not to the root of the repository, which sounds reasonable to me.

**Review by TingPing — COMMENTED** [review]
CMake changes LGTM

**Pull request merged**
